### ISSUE TITLE
Blazegraph fork?

### ISSUE THREAD

**Issue by berezovskyi** [blazegraph/database#86]
Now that it seems Amazon has acquired Blazegraph (https://www.trademarkia.com/blazegraph-86498414.html etc), does it make sense for the OSS community to fork Blazegraph and restart the development (at least processing the PRs)?

**Comment by ianwdunlop** [comment]
Another goal of a forked/reborn blazegraph would be to migrate the sesame dependency to org.eclipse.rdf4j.

**Comment by ianwdunlop** [comment]
Anyone still interested in forking blazegraph and trying to keep it alive in the OSS community? Maybe someone already has. If so where is the project? Maybe a possible strategy is to fork and try to build from master and the latest unreleased branch (maybe 2.2.0). I've tried building but there are some test failures. Has anyone else tried building from source?

**Comment by beebs-systap** [comment]
@ianwdunlop @berezovskyi We've had some recent discussions on this and there seems to be community interest, which we're working to gather together. Feel free to update this interest for folks that would be interested.

@smalyshev @thompsonbry

**Comment by ianwdunlop** [comment]
@nyurik That's a fair proposal. It would be good to hear more from @beebs-systap and folks about what their intentions are - whether they would like to keep the original blazegraph repo going or if it is time to start a new project based on the existing code.

**Comment by thompsonbry** [comment]
There have been some discussions about moving the project to wikidata /
wikimedia and some strong expressions of interest for that from the their
community.  Adding Stas and Beebs..

Bryan

**Comment by nyurik** [comment]
I think @smalyshev is doing an amazing job with various Blazegraph improvements, and certainly should be one of the key people driving this project forward, but I do not think Blazegraph fork should be under Wikimedia's umbrella, using less convenient or familiar tooling (gerrit/phabricator), or managed by WMF management. Rather, I think it should remain in GitHub e.g. a separate organization as a typical open source project, with a low barrier of entry, a wide community of contributing and reviewing developers, and WMF developers actively participating.

**Comment by thompsonbry** [comment]
I suspect that the project could stay where it is in terms of tooling.  The
question is more about the responsibility for producing releases.  Brad has
discussed some options for providing CI resources.

It would be great to get a sense of how many people would be interested in
leaning in to help wikimedia support the open source blazegraph platform.
That might help them make this transition happen.

Thanks,
Bryan

**Comment by turbolent** [comment]
In https://github.com/turbolent/embergraph I've took 2.1.5-RC and started renaming the code to Embergraph (to avoid inflicting on trademarks), cleaning up the code base, and updating it (Java 8).

I plan on mostly updating the SPARQL server (probably replacing Jetty and servlets with something else) and will try to update dependencies (e.g. migrate to RDF4J).

**Comment by thompsonbry** [comment]
Awesome for getting involved.  However, you might consider just issuing
CRs.  Stas should have access to a CI setup and he and/or Brad can cut
releases.

Bryan

**Comment by thompsonbry** [comment]
CR == pull request

**Comment by beebs-systap** [comment]
@turbolent Would you be amenable to setting up a CLA to push the changes back into the core Blazegraph?

**Comment by ianwdunlop** [comment]
2 quick comments. Why do we need a CLA to push changes (via pull request I assume) back into the core blazegraph repo? Is this a different repo from this one? Isn't all this covered by the current licence (GPL 2)?
From recent comments it seems that you would like to restart blazegraph development using this repo rather than fork off to anywhere else. I like that idea.

**Comment by nyurik** [comment]
> In https://github.com/turbolent/embergraph I've took 2.1.5-RC and started renaming the code to Embergraph (to avoid inflicting on trademarks), cleaning up the code base, and updating it (Java 8).
> 
> I plan on mostly updating the SPARQL server (probably replacing Jetty and servlets with something else) and will try to update dependencies (e.g. migrate to RDF4J).

I do like the Embergraph name, but there does seem to be [another project](https://github.com/stingerlabs/ember-graph) by that name. Not sure if that's a problem.

I am not sure what you meant by `renaming the code to Embergraph` - IANAL, but it seems [MariaDB](https://github.com/MariaDB/server/search?q=mysql&unscoped_q=mysql) who went through the similar process did not modified the actual code - as that would make the code incompatible with extensions/external modifications.  There is a difference between classnames that use `Blazegraph` or `Bigdata` internally, and the branding of the product.  I suspect we only need to modify the later.

**Comment by beebs-systap** [comment]
@ianwdunlop @nyurik Yes, you can definitely fork it under GPLv2 into a separate project. However, we would like to see if there's interest in getting a set of interested parties to collaborate on restarting development in this github. WMF/Wikidata would definitely have a role as a significant stakeholder, but is not the only one. The CLA is needed to make sure that the contributions can be accepted under GPLv2, but it can be a fairly lightweight process. Would you be amenable to setting up a quick chat to discuss in more details?

**Comment by nyurik** [comment]
@beebs-systap I obviously would prefer not to fork if possible, and sure, ping me at my firstandlastname at gmail :) I was under the impression that now that the trademark belongs to Amazon, no one outside of Amazon can use it, hence the need for a fork.  Another (possibly minor or irrelevant?) concern is re-licensing - e.g. if a person contributes code to this repo under GPL2, can Amazon re-license that contribution under a proprietary license?

**Comment by ianwdunlop** [comment]
I agree with @nyurik that keeping it in the original repo is preferred over a fork. If a CLA is necessary to contribute under gpl2 then that seems ok. I imagine what contributors need assurance over is that any code they commit remains open source forever. I guess we would have to clarify any trademark issues since that could cause problems for some people. Ping me at gmail for a lightweight discussion. One thought that did occur to me is what branch development should be focussed on? master or the latest unreleased branch or another? What would be the priorities for a re-invigorated blazegraph? I think it would be getting dependencies up to the latest rdf4j etc rather than focussing on new features.

**Comment by beebs-systap** [comment]
@ianwdunlop @nyurik Sounds workable. Yes, everything would be open source under GPLv2 and the CLA is a fairly standard process (it has always been in-place for Bigdata/Blazegraph). I think there's two priorities; updating the 2.1.5 branch as needed and picking up the 2.2.0 branch for future work.

Let me setup a quick meeting/discussion offline with you and we can chat through some of the details.

**Comment by thompsonbry** [comment]
We had some discussions at ISWC (last fall) with wikidata and some others
on priorities.  Priorities should emerge out of the community, but clearly
updating the dependencies would be high on the list based on previous
discussions.

Bryan

**Comment by berezovskyi** [comment]
I am not sure if depending on Amazon's goodwill regarding the trademark is a good idea. Also, CLA likely means a copyright assignment process which would allow Amazon to use GPLed contributions in Neptune without contributing any patches back to Blazegraph under GPL, which might not be in the spirit of GPL.

But I understand that Wikimedia and a few others are the major users of Blazegraph and we should hear their opinion. From a cursory look, seems like @turbolent and @ianwdunlop have the most active forks and the former is free from trademark use.

**Comment by thompsonbry** [comment]
The CLA has been in place since 2006. We always accepted contributions
under a CLA in order to help protect everyone’s rights. Of course, there
are plenty of opinions about open source, but this is how blazegraph
contributions were managed from the project inception. It is nothing new
with Amazon.

Bryan

**Comment by akuckartz** [comment]
Is that CLA published anywhere?

**Comment by thompsonbry** [comment]
I don't see it up online off hand.  It is referenced at [1,2].   Brad would
be able to provide you with a current version.

Thanks,
Bryan

[1] https://wiki.blazegraph.com/wiki/index.php/Contributors
[2] https://www.blazegraph.com/develop/

**Comment by berezovskyi** [comment]
These links may be of use to see how it looked in the past:

https://web.archive.org/web/20140103220007/http://www.systap.com/BigdataIndividualCA.pdf

and 

https://web.archive.org/web/20140103215942/http://www.systap.com/BigdataCorporateCA.pdf

**Comment by elft3r** [comment]
Hi everybody,
I'm currently in the process of evaluating different graph databases and one of them is blazegraph. It looks really promising, and we are thinking about choosing it for our solution.

However, the uncertainty of the future of this project, lets us still hesitate whether we should adopt it or not. Are there any updates about the plans of this project?

Thanks for your help,
Jochen

**Comment by beebs-systap** [comment]
@elft3r Wikidata is running it in production and there's interest in continuing to move it forward.

**Comment by elft3r** [comment]
@beebs-systap Thanks for your quick reply. Are there already plans laid out on how it will move forward?

**Comment by beebs-systap** [comment]
@elft3r There's ongoing discussion and a general outline, which is in process.

**Comment by berezovskyi** [comment]
I am ok closing the issue given that there are new releases.

**Comment by josn0w** [comment]
Hi all, what is the current status?
There was no new release for almost one year ...

**Comment by josn0w** [comment]
Well, as there is nothing new, should the project considerd dead?

> I don't see it up online off hand. It is referenced at [1,2]. Brad would be able to provide you with a current version. Thanks, Bryan [1] https://wiki.blazegraph.com/wiki/index.php/Contributors [2] https://www.blazegraph.com/develop/

And those links are not working anymore...

> In https://github.com/turbolent/embergraph I've took 2.1.5-RC and started renaming the code to Embergraph (to avoid inflicting on trademarks), cleaning up the code base, and updating it (Java 8).

This fork is also not maintained anymore...

I would really like to help here.

**Comment by berezovskyi** [comment]
Seems like teaming up with Wikimedia is the best course of action, their fork is very much alive:

- https://gerrit.wikimedia.org/r/q/project:wikidata%252Fquery%252Fblazegraph
- https://phabricator.wikimedia.org/search/query/jVcJQvQHKgRG/#R

**Comment by josn0w** [comment]
So what should be the plan here?

**Comment by berezovskyi** [comment]
TBH, I only met one person from Wikimedia DE / Wikidata at a conference a few years ago who was not directly involved with Wikidata infra. We need to find a point of contact at the Wikidata infra team first to see if they are ready to lead the fork development (at least to package and publish it). Of course, would have been better if new maintainers could simply get write access to this org instead of forking...

**Comment by josn0w** [comment]
I see, who has write access?

AFAIK the wmf fork is not really maintained as well..

**Comment by berezovskyi** [comment]
@beebs-systap and @thompsonbry I would presume.

**Comment by josn0w** [comment]
Ok, i saw the jira was last maintained by @beebs-systap  and @igor-kim (Jan. 2020).

But i dont know if one should work on the Tickets as, they are not maintained as well..

**Comment by josn0w** [comment]
FYI: I got a simple CI Pipeline working (build the jar + Docker Image)

**Comment by beebs-systap** [comment]
@josn0w @berezovskyi I'd prefer to avoid a fork too. If you want to contribute, let's get a CLA in place and do it from within the current project. I'll reach out separately to discuss.

We'd also be interested in any others that also want to contribute.

**Comment by josn0w** [comment]
@beebs-systap I would like to contribute here, if it is possible.
It would be great if you can update on the current status of the Project and the plan for the future.

**Comment by beebs-systap** [comment]
@josn0w Great. Send me an email `beebs@amazon.com` RE: the CLA. The status of the project and the plan for the future depends on the contributors. We're happy to continue to support.

**Comment by beebs-systap** [comment]
@josn0w Great. Send me an email `beebs [at] amazon.com` RE: the CLA. The status of the project and the plan for the future depends on the contributors. We're happy to continue to support.

**Comment by jeffreyschultz** [comment]
any update on this discussion?

**Comment by berezovskyi** [comment]
@jeffreyschultz I am using Jena/Fuseki and the built-in SWI-Prolog RDF store for most of my project work now. Not sure if @ben-j-herbertz got push/releng/issue triage rights to the repo. 
I guess housekeeping like adding CI and updating CVE-affected dependencies would be the first step. Ensuring https://w3c.github.io/rdf-tests/ are covered would be another, though fixing issues like https://github.com/blazegraph/database/issues/234 would require some non-trivial commitment.

Wikimedia is evaluating migration off Blazegraph: https://phabricator.wikimedia.org/T206560
